iscan/static/iscan/query-detail/query-detail.js

```diff
@@ -31,415 +31,485 @@ angular.module('queryDetail', [
+            for (i = 0; i < $scope.hierarchy.speaker_properties.length; i++) {
+                prop = $scope.hierarchy.speaker_properties[i][0];
+                if ($scope.properties.speaker.indexOf(prop) === -1 && prop !== 'id') {
+                    $scope.properties.speaker.push(prop);
+                    $scope.propertyValues.speaker[prop] = $scope.speaker[prop];
                 }
-            );
-                })
-                    .catch(function (data) {
-                        if (data.status === 423) {
-                            $scope.utterance = {};
-                            $scope.headline = 'Could not load utterance';
-                        }
-                    });
-            };
+            }
+        }
+    };
 
 
+    $scope.deleteAnnotation = function (id) {
+        Annotations.delete($stateParams.corpus_id, id).then(function (res) {
+            for (i = 0; i < $scope.annotations.length; i++) {
+                $scope.currentAnnotations[$scope.annotations[i].label] = $scope.currentAnnotations[$scope.annotations[i].label].filter(function (annotation) {
+                    return annotation.id !== id
+                });
+            }
+        });
+    };
+
+    $scope.updateAnnotation = function (annotation) {
+        Annotations.update($stateParams.corpus_id, annotation.id, annotation).then();
+    };
 
-        djangoAuth.authenticationStatus(true).then(function () {
+    $scope.addAnnotation = function (annotation) {
+        var data = {};
+        data.annotation_type = $scope.query.annotation_type.toLowerCase();
+        data.annotation_id = $scope.selectedAnnotation.id;
+        data.subannotation_type = annotation.label;
+        data.subannotation = $scope.newAnnotation[annotation.label];
 
-            Users.current_user().then(function (res) {
-                $scope.user = res.data;
-                $scope.refreshPermissions();
+        if (annotation.save_user) {
+            data.subannotation.user = $scope.user.username;
+        }
+        Annotations.create($stateParams.corpus_id, data).then(function (res) {
+            $scope.currentAnnotations[annotation.label].push(res.data);
+        })
+    };
+
+
+    $scope.runQuery = function () {
+        if ($scope.user.is_superuser) {
+            $scope.can_listen = true;
+            $scope.can_edit = true;
+            $scope.can_view_annotations = true;
+            $scope.can_annotate = true;
+
+        }
+        else {
+            $scope.can_listen = false;
+            $scope.can_edit = false;
+            $scope.can_view_annotations = false;
+            $scope.can_annotate = false;
+            for (i = 0; i < $scope.user.corpus_permissions.length; i++) {
+                if ($scope.user.corpus_permissions[i].corpus == $stateParams.corpus_id) {
+                    $scope.can_listen = $scope.user.corpus_permissions[i].can_listen;
+                    $scope.can_edit = $scope.user.corpus_permissions[i].can_edit;
```

Comment:
Ok, wait, there's this here now in addition to the code above in `refreshPermissions`.  This block about permissions might be better moved to the `refreshPermissions` function?